<?xml version="1.0" encoding="utf-8"?>
<robot name="mc_robo">

  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="blue">
    <color rgba="0.203125 0.23828125 0.28515625 1.0"/>
  </material>
  <material name="green">
    <color rgba="0.0 0.8 0.0 1.0"/>
  </material>
  <material name="grey">
    <color rgba="0.2 0.2 0.2 1.0"/>
  </material>
  <material name="orange">
    <color rgba="1.0 0.423529411765 0.0392156862745 1.0"/>
  </material>
  <material name="brown">
    <color rgba="0.870588235294 0.811764705882 0.764705882353 1.0"/>
  </material>
  <material name="red">
    <color rgba="0.80078125 0.12890625 0.1328125 1.0"/>
  </material>
  <material name="white">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>

  <gazebo reference="link_chassis">
    <material>Gazebo/Orange</material>
  </gazebo>
  <gazebo reference="link_left_wheel">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="link_right_wheel">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="beltwagon">
    <material>Gazebo/white</material>
  </gazebo>
  <gazebo reference="rear_boom">
    <material>Gazebo/brown</material>
  </gazebo>
  <gazebo reference="front_boom">
    <material>Gazebo/brown</material>
  </gazebo>

  <!-- making real laser scanner in gazebo-->
  <gazebo reference="sensor_laser">
    <sensor type="ray" name="head_hokuyo_sensor">
      <pose>0 0 0 0 0 0</pose>
      <visualize>false</visualize>
      <update_rate>20</update_rate>
      <ray>
        <scan>
          <horizontal>
            <samples>720</samples>
            <resolution>1</resolution>
            <min_angle>-1.570796</min_angle>
            <max_angle>1.570796</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.10</min>
          <max>10.0</max>
          <resolution>0.01</resolution>
        </range>
        <noise>
          <type>gaussian</type>
          <mean>0.0</mean>
          <stddev>0.01</stddev>
        </noise>
      </ray>
      <plugin name="gazebo_ros_head_hokuyo_controller" filename="libgazebo_ros_laser.so">
        <topicName>/mc_robo_laser_scan</topicName>
        <frameName>sensor_laser</frameName>
      </plugin>
    </sensor>
  </gazebo>

  <gazebo>
    <plugin filename="libgazebo_ros_diff_drive.so" name="differential_drive_controller">
      <legacyMode>false</legacyMode>
      <alwaysOn>true</alwaysOn>
      <updateRate>20</updateRate>
      <leftJoint>joint_left_wheel</leftJoint>
      <rightJoint>joint_right_wheel</rightJoint>
      <wheelSeparation>0.2</wheelSeparation>
      <wheelDiameter>0.2</wheelDiameter>
      <torque>0.1</torque>
      <commandTopic>cmd_vel</commandTopic>
      <odometryTopic>odom</odometryTopic>
      <odometryFrame>odom</odometryFrame>
      <robotBaseFrame>link_chassis</robotBaseFrame>
    </plugin>
  </gazebo>

  <link name="link_chassis"> <!-- consider as a base_link -->
    <!-- pose and inertial -->
    <pose>0 0 0.1 0 0 0</pose>
    <inertial>
      <mass value="50"/>
      <origin rpy="0 0 0" xyz="0 0 0.1"/>
      <inertia ixx="68.7083333333333" ixy="0" ixz="0" iyy="104.166666666666" iyz="0" 		 	        izz="39.541666666666"/>
    </inertial>
    <!-- body -->
    <collision name="collision_chassis">
      <geometry>
        <box size="0.3 0.3 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.3 0.3 0.05"/>
      </geometry>
      <material name="orange"/>
    </visual>

    <!-- caster front -->
    <!--<collision name="caster_front_collision">
      <origin rpy=" 0 0 0" xyz="0.35 0 -0.05"/>
      <geometry>
        <sphere radius="0.05"/>
      </geometry>
      <surface>
        <friction>
          <ode>
            <mu>0</mu>
            <mu2>0</mu2>
            <slip1>1.0</slip1>
            <slip2>1.0</slip2>
          </ode>
        </friction>
      </surface>
    </collision>
    <visual name="caster_front_visual">
      <origin rpy=" 0 0 0" xyz="0.2 0 -0.05"/>
      <geometry>
        <sphere radius="0.05"/>
      </geometry>
    </visual>
  </link>-->

    <!-- caster back -->     <!-- passive joint-->
   <inertial>    
      <mass value="1"/>
      <origin rpy=" 0 0 0" xyz="-0.15 0 -0.025"/>
      <inertia ixx="0.009" ixy="0" ixz="0" iyy="0.009" iyz="0" 		 	        		izz="0.009"/>
    </inertial>
    <collision name="caster_back_collision">
      <origin rpy=" 0 0 0" xyz="-0.12 0 -0.025"/>
       <geometry>
        <sphere radius="0.015"/>
      </geometry>
      <surface>
        <friction>
          <ode>
            <mu>0</mu>
            <mu2>0</mu2>
            <slip1>1.0</slip1>
            <slip2>1.0</slip2>
          </ode>
        </friction>
      </surface>
    </collision>
    <visual name="caster_front_visual">
      <origin rpy=" 0 0 0" xyz="-0.12 0 -0.025"/>
      <geometry>
        <sphere radius="0.015"/>
      </geometry>
    </visual>
  </link>

    <!-- right_wheel -->
  <link name="link_right_wheel">
    <inertial>
       <mass value="2"/>
       <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
       <inertia ixx="0.1066666666" ixy="0" ixz="0" iyy="0.1066666666" iyz="0" 			   izz="0.16"/>
    </inertial>
    <collision name="link_right_wheel_collision">
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.04" radius="0.04"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
       <cylinder length="0.04" radius="0.04"/>
      </geometry>
    </visual>
  </link>

  <joint name="joint_right_wheel" type="continuous">
     <origin rpy="0 0 0" xyz="0.05 0.15 0"/>
     <child link="link_right_wheel"/>
     <parent link="link_chassis"/>
     <axis rpy="0 0 0" xyz="0 1 0"/>
     <limit effort="10" velocity="1000"/>
     <joint_properties damping="0.0" friction="0.0"/>
  </joint>

   <!-- left_wheel -->
  <link name="link_left_wheel">
    <inertial>
       <mass value="2"/>
       <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
       <inertia ixx="0.1066666666" ixy="0" ixz="0" iyy="0.1066666666" iyz="0" 			   izz="0.16"/>
    </inertial>
    <collision name="link_left_wheel_collision">
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.04" radius="0.04"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
       <cylinder length="0.04" radius="0.04"/>
      </geometry>
    </visual>
  </link>

  <joint name="joint_left_wheel" type="continuous">
     <origin rpy="0 0 0" xyz="0.05 -0.15 0"/>
     <child link="link_left_wheel"/>
     <parent link="link_chassis"/>
     <axis rpy="0 0 0" xyz="0 1 0"/>
     <limit effort="10" velocity="1000"/>
     <joint_properties damping="0.0" friction="0.0"/>
  </joint>

   <!-- mobile_crusher -->
  <link name="mobile_crusher">
   <inertial>
      <mass value="5"/>
      <origin rpy="0 0 0" xyz="0 0 0.06"/>
      <inertia ixx="0.455" ixy="0" ixz="0" iyy="0.455" iyz="0" 		 	        		izz="0.75"/>
    </inertial>
    <collision name="mobile_crusher">
      <geometry>
        <cylinder length="0.08" radius="0.1"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.08" radius="0.1"/>
      </geometry>
      <material name="white"/>
    </visual>
  </link>

  <joint name="chassis_to_beltwagon" type="fixed">
     <origin rpy="0 0 0" xyz="0 0 0.06"/>
     <child link="mobile_crusher"/>
     <parent link="link_chassis"/>
     <axis rpy="0 0 0" xyz="0 0 1"/>
     <limit effort="10" lower="-1.5" upper="1.5" velocity="10"/>
  </joint>

   <!-- rear_boom -->
  <link name="rear_boom">
    <inertial>
     <mass value="2"/>
     <origin rpy="0 0 0" xyz="0 -0.1 0.09"/> <!-- centre of mass-->
     <inertia ixx="0.00708333333" ixy="0" ixz="0" iyy="0.00708333333" iyz="0" 		 	        izz="8.333333333e-4"/>
    </inertial>
    <collision name="rear_boom">
      <geometry>
        <box size="0.05 0.2 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0.1 0"/> <!-- tf axis-->
      <geometry>
        <box size="0.05 0.2 0.05"/>
      </geometry>
     </visual>
  </link>

  <joint name="beltwagon_to_rear_boom" type="revolute">
     <origin rpy="0 0 0" xyz="0 0 0.06"/> <!-- joint placed between two links-->
     <child link="rear_boom"/>
     <parent link="mobile_crusher"/>
     <axis rpy="0 0 0" xyz="1 0 0"/> <!-- rotation around x axis -->
     <limit effort="1000" lower="-0.25" upper="0.25" velocity="1000"/>
  </joint>

  <!-- add laser sensor --> 
  <link name="sensor_laser">
    <inertial>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <mass value="1" />
        <inertia  ixx="0.001458333" ixy = "0" ixz = "0"
                  iyy="0.001458333" iyz = "0"
                  izz="0.00125" />
    </inertial>

    <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
        <cylinder radius="0.02" length="0.03"/>
        </geometry>
        <material name="black" />
    </visual>

    <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
        <cylinder radius="0.02" length="0.03"/>
        </geometry>
    </collision>
  </link>

  <joint name="joint_sensor_laser" type="fixed">
    <origin rpy="0 0 0" xyz="0.06 0 0.05"/>
    <parent link="mobile_crusher"/>
    <child link="sensor_laser"/>
  </joint>	

 </robot>
